### ISSUE TITLE
Simulate Call as ContractID from WalletUnlocked?

### ISSUE THREAD

**Issue by diyahir** [FuelLabs/fuels-rs#952]
I have a contract A that is restricted to be called by another contract B which is of type ‘ContractID’.

But I want to test A quickly without spinning up both A and B, currently, I am only aware of making calls as ‘Identity::Address’.

Is there a way to pretend to be a ContractID in a rust integration test instead of just a wallet identity?

This will greatly simplify the testing process without needing to get B to the proper state to call A.

Ethereum doesn’t make a strong distinction between contracts and eoa’s wrt to testing, so I’m wondering if Fuel has plans to do this as well

```
storage {
    adminID: Identity = Identity(...),
    adminCID: ContractId = ContractId(...),
}
```

In this case it's easier to run an integration test if the admin is an identity since you can use wallet unlocked class to test it easily, but the production it should be a ContractId since it's a contract

Is the recommendation to always use Identity for easier testing? that would make using the contract's abi more complicated since you would need to cast the Identity to a ContractId 

```
let admin_contract = abi(Admin, storage.adminCID.value);
```

**Comment by MujkicA** [comment]
`msg_sender` which is used on the sway side to retrieve the calling identity checks the context from which the call is being made. If it is internal (from another contract), it returns the calling contracts id, otherwise it looks at the tx inputs and if all coins/messages have the same owner, it returns that address.

AFAIK, there is no way to trick the VM into thinking that an external call is made from an internal context, we only interact with it via the node client after all. Even if there was, I would be wary of the consequences this may have on execution.

Supporting this would require us to modify the inputs of the call transaction such that their owner field is the address we are trying to mock. Also utxo validation would need to be set to false in the node config.

tldr: I think this would be possible, just a matter of priority.

About `Identity`, I am no position to make recommendations since sway is also involved but since `msg_sender` returns `Identity` it seems like you would need a cast at some point either way.

**Comment by diyahir** [comment]
It's not clear to me why make the distinction at all, it seems a bit like technical debt.

This whole thing about Address, ContractId, Identity, and AssetId adds a layer of complexity that I don't think adds much value to the sway stack. It just adds a type-casting nightmare from my experience.

When doing integration tests, without the ability to pretend to be a contract it means there's an immense amount of overhead just to get the conditions properly in one contract to ultimately make the desired call to the second contract.

**Comment by meirbank** [comment]
@MujkicA We have a lot of contracts in our protocol, more than most (all?) others built in Sway. So this problem is very acute for us in complicating testing. If we just switch from ContractId to Identity we would be able to avoid this. Would you recommend this?

**Comment by MujkicA** [comment]
> @MujkicA We have a lot of contracts in our protocol, more than most (all?) others built in Sway. So this problem is very acute for us in complicating testing. If we just switch from ContractId to Identity we would be able to avoid this. Would you recommend this?

You could use Identity and set your adminId via configurable constants. This will let you deploy your contracts in a test environment with the test wallet as admin, while the prod environment could be setup with a contract as admin.

**Comment by diyahir** [comment]
> While it's true that Address, ContractId, Identity, and AssetId are all represented as 32-byte arrays, their usage contexts differ. I believe making this distinction is a good practice as it let's the type checker keep you safe from e.g. accidentally passing an AssetId where an Address is expected.

I understand the argument for Identity/Contract/Address vs AssetId, but I would still argue that the distinction between Identity, Contract, and Address is not a useful distinction: Imagine you need to go from a single signer admin, to a multisig (if you didn't have the foresight you cannot do this). Or you defi application used address everywhere instead of identity then their protocol is no longer composable with other contracts, since a contractID is not an Address and will break their logic. 

You can check out swaylend's current contract. Thinking about this bug, I looked through some contracts and here you can see they run into this mistake:

````
 #[storage(read)]
    fn get_user_collateral(address: Address, asset: ContractId) -> u64;

 #[storage(read)]
    fn is_liquidatable(account: Address) -> bool;
````

They incorrectly assume anyone using the platform is a user, not a contract, so now they are no longer composable with other contracts. This is a critical mistake, and one I can see many people running into. By creating this distinction, sway actually introduces a new class of un-composability bugs.

https://github.com/compolabs/sway-lend/blob/master/contracts/market/src/main.sw

**Comment by meirbank** [comment]
> > @MujkicA We have a lot of contracts in our protocol, more than most (all?) others built in Sway. So this problem is very acute for us in complicating testing. If we just switch from ContractId to Identity we would be able to avoid this. Would you recommend this?
> 
> You could use Identity and set your adminId via configurable constants. This will let you deploy your contracts in a test environment with the test wallet as admin, while the prod environment could be setup with a contract as admin.

Thanks for the suggestion, this sounds good but the issue is that everything that the constant touches will need to know if it is an Identity or a ContractId. 

For instance we would have to switch requires everywhere from
`require(msg_sender().unwrap() == Identity::ContractId(storage.stability_pool_contract), "BorrowOperations: Caller is not Stability Pool");`
to
`require(msg_sender().unwrap() == storage.stability_pool_contract, "BorrowOperations: Caller is not Stability Pool");`
and then back after testing with Identity

**Comment by MujkicA** [comment]
> > > @MujkicA We have a lot of contracts in our protocol, more than most (all?) others built in Sway. So this problem is very acute for us in complicating testing. If we just switch from ContractId to Identity we would be able to avoid this. Would you recommend this?
> > 
> > 
> > You could use Identity and set your adminId via configurable constants. This will let you deploy your contracts in a test environment with the test wallet as admin, while the prod environment could be setup with a contract as admin.
> 
> Thanks for the suggestion, this sounds good but the issue is that everything that the constant touches will need to know if it is an Identity or a ContractId.
> 
> For instance we would have to switch requires everywhere from `require(msg_sender().unwrap() == Identity::ContractId(storage.stability_pool_contract), "BorrowOperations: Caller is not Stability Pool");` to `require(msg_sender().unwrap() == storage.stability_pool_contract, "BorrowOperations: Caller is not Stability Pool");` and then back after testing with Identity

I don't quite follow, why would you need to change it back? The idea is that your contract is agnostic w.r.t. the identity type so that you can use one type while testing and another in production. Unless you want your prod version to only permit contracts?

**Comment by MujkicA** [comment]
> > While it's true that Address, ContractId, Identity, and AssetId are all represented as 32-byte arrays, their usage contexts differ. I believe making this distinction is a good practice as it let's the type checker keep you safe from e.g. accidentally passing an AssetId where an Address is expected.
> 
> I understand the argument for Identity/Contract/Address vs AssetId, but I would still argue that the distinction between Identity, Contract, and Address is not a useful distinction: Imagine you need to go from a single signer admin, to a multisig (if you didn't have the foresight you cannot do this). Or you defi application used address everywhere instead of identity then their protocol is no longer composable with other contracts, since a contractID is not an Address and will break their logic.
> 
> You can check out swaylend's current contract. Thinking about this bug, I looked through some contracts and here you can see they run into this mistake:
> 
> ```
>  #[storage(read)]
>     fn get_user_collateral(address: Address, asset: ContractId) -> u64;
> 
>  #[storage(read)]
>     fn is_liquidatable(account: Address) -> bool;
> ```
> 
> They incorrectly assume anyone using the platform is a user, not a contract, so now they are no longer composable with other contracts. This is a critical mistake, and one I can see many people running into. By creating this distinction, sway actually introduces a new class of un-composability bugs.
> 
> https://github.com/compolabs/sway-lend/blob/master/contracts/market/src/main.sw

That seems like a valid point. I see how people coming from different ecosystems could easily make this mistake.
Still I keep thinking that there is some benefit there too as it enables you to to fine tune how composable you want to be by opting for `Identity` or a specific type. 
The sway docs could definitely benefit from being more explicit about the consequences of this choice.

**Comment by diyahir** [comment]
I would really push back against calling it a 'feature' not a 'bug', to make it concrete, swaylend is a compound implementation, and many protocols hook into compound as smart contracts. There are very few instances, if any, in which you would want to limit your user base to just Addresses.

If one of the most prominent fuel projects made this mistake, then there's little hope for other teams not as familiar with the stack. I recommend removing this Identity, Address, and Contract distinction as it will only cause developers to make incompatibility mistakes. 

My biggest issue with this whole thing is that it separates Multisigs from Address, so if you ever want to manage a position as a DAO, you NEED the contract to allow Identity. In any case of a multi-sig/DAO, you need to pray and hope that the devs considered that use case. In basically all cases, in DeFi, in social graphs, in NFT marketplaces, sometimes you want to do something as a community, and you need to have this capability supported by default.

I think this is a CRITICAL error, placing multi sigs as second-class citizens is not the right direction. I don't think the full implications of this distinction have been thoroughly explored. I suspect it's something that can only have been revealed later in development so I don't fault anyone for not seeing its implications.

**Comment by MujkicA** [comment]
You make a valid point about this being a huge risk factor. However, my opinion remains that for a smart contract, any choice of permitting additional functionality/use-cases should always be deliberate instead of implied. I think this is better addressed trough proper documentation. 

That being said, this is the Rust SDK repo, we don't drive Sway design decisions so I've opened an issue here:
https://github.com/FuelLabs/sway/issues/4569

Please feel free to share further thoughts on this matter there.

This thread should focus on SDK support for mocking calls from different sources.
